accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -0,0 +1,536 @@
+variables, static fields, instance fields, and local variables is extended to
+allow any declaration to include the modifer `late`.
+
+The modifier `required` is added as a built-in identifier. The grammar of
+function types is extended to allow any named parameter declaration to be
+prefixed by the `required` modifier (e.g. `int Function(int, {int?  y, required
+int z})`. 
+
+The grammar of selectors is extended to allow null-aware subscripting using the
+syntax `e1?.[e2]` which evaluates to `null` if `e1` evaluates to `null` and
+otherwise evaluates as `e1[e2]`.
+
+
+### Grammatical ambiguities and clarifications.
+
+#### Nested nullable types
+
+The grammar for types does not allow multiple successive `?` operators on a
+type.  That is, the grammar for types is nominally equivalent to:
+
+```
+type' ::= functionType
+          | qualified typeArguments? 
+
+type ::= type' `?`?
+```
+
+#### Conditional expression ambiguities
+
+Conditional expressions inside of braces are ambiguous between sets and maps.
+That is, `{ a as bool ? - 3 : 3 }` can be parsed as a set literal `{ (a as bool)
+? - 3 : 3 }` or as a map literal `{ (a as bool ?) - 3 : 3 }`.  Parsers will
+prefer the former parse over the latter.
+
+The same is true for `{ a is int ? - 3 : 3 }`.
+
+The same is true for `{ int ? - 3 : 3 }` if we allow this.
+
+
+## Static semantics
+
+### Legacy types
+
+The internal representation of types is extended with a type `T*` for every type
+`T` to represent legacy pre-NNBD types.  This is discussed further in the legacy
+library section below.
+
+### Static errors
+
+We say that a type `T` is **nullable** if `Null <: T`.  This is equivalent to
+the syntactic criterion that `T` is any of:
+  - `Null`
+  - `S?` for some `S`
+  - `S*` for some `S` where `S` is nullable
+  - `FutureOr<S>` for some `S` where `S` is nullable
+  - `dynamic`
+  - `void`
+
+We say that a type `T` is **non-nullable** if `T <: Object`.  This is equivalent
+to the syntactic criterion that `T` is any of:
+  - `Object`, `int`, `bool`, `Never`, `Function`
+  - Any function type
+  - Any class type or generic class type
+  - `FutureOr<S>` where `S` is non-nullable
+  - `X extends S` where `S` is non-nullable
+  - `X & S` where `S` is non-nullable
+
+Note that there are types which are neither nullable nor non-nullable.  For
+example `X extends T` where `T` is nullable is neither nullable nor
+non-nullable.
+
+We say that a type `T` is **potentially nullable** if `T` is not non-nullable.
+Note that this is different from saying that `T` is nullable.  For example, a
+type variable `X extends Object?` is a type which is potentially nullable but
+not nullable.  Note that `T*` is always potentially nullable by this definition.
+
+We say that a type `T` is **potentially non-nullable** if `T` is not nullable.
+Note that this is different from saying that `T` is non-nullable.  For example,
+a type variable `X extends Object?` is a type which is potentially non-nullable
```

Comment:
Shouldn't it be `X extends Object` instead of  `X extends Object?`?

Reply:
No.  Note that `Object` is not nullable.

`X extends Object` **is** potentially non-nullable, since it is not nullable, but it is **also** non-nullable, since it is a subtype of `Object`.  

`X extends Object?` on the other hand is potentially non-nullable, since it is not nullable (`Null` is not a subtype of it) , but it is not *known* to be non-nullable, since it could be instantiated with a nullable type.

That may all seem confusing, but basically:

nullable type -> definitely could get Null
  - e.g. int?
non-nullable type -> definitely will never get Null
  - e.g. int
potentially nullable type -> might or might not get "instantiated" such that it becomes nullable
  - e.g. X extends Object?
  - e.g. int*
potentially non-nullable type -> might or might not get "instantiated such it becomes non-nullable. 
  - e.g. X extends Object?
  - e.g. int*

Reply:
Thanks for the explanation.